src/main/java/com/github/mizool/jediscache/JedisCachingProvider.java

```diff
@@ -0,0 +1,211 @@
+/**
+ * Copyright 2017-2018 incub8 Software Labs GmbH
+ * Copyright 2017-2018 protel Hotelsoftware GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.github.mizool.jediscache;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.HashMap;
+import java.util.Properties;
+import java.util.WeakHashMap;
+
+import javax.cache.CacheException;
+import javax.cache.CacheManager;
+import javax.cache.configuration.OptionalFeature;
+import javax.cache.spi.CachingProvider;
+
+import lombok.extern.slf4j.Slf4j;
+
+import org.kohsuke.MetaInfServices;
+
+import redis.clients.jedis.JedisPool;
+import redis.clients.jedis.JedisPoolConfig;
+
+@Slf4j
+@MetaInfServices(CachingProvider.class)
+public class JedisCachingProvider implements CachingProvider
+{
+    private final JedisPool jedisPool;
+    private WeakHashMap<ClassLoader, HashMap<URI, CacheManager>> cacheManagersByClassLoader;
+
+    public JedisCachingProvider()
+    {
+        String host = System.getProperty("jediscache.host", "127.0.0.1");
+        log.info("JedisCache active, connecting to {}", host);
+        this.jedisPool = new JedisPool(new JedisPoolConfig(), host);
+        this.cacheManagersByClassLoader = new WeakHashMap<>();
+    }
+
+    @Override
+    public synchronized CacheManager getCacheManager(URI uri, ClassLoader classLoader, Properties properties)
+    {
+        if (uri == null)
+        {
+            uri = getDefaultURI();
+        }
+        if (classLoader == null)
+        {
+            classLoader = getDefaultClassLoader();
+        }
+        if (properties == null)
+        {
+            properties = new Properties();
+        }
+
+        HashMap<URI, CacheManager> cacheManagersByURI = cacheManagersByClassLoader.get(classLoader);
+
+        if (cacheManagersByURI == null)
+        {
+            cacheManagersByURI = new HashMap<>();
+        }
+
+        CacheManager cacheManager = cacheManagersByURI.get(uri);
+
+        if (cacheManager == null)
+        {
+            cacheManager = new JedisCacheManager(this, uri, classLoader, properties, jedisPool);
+            cacheManagersByURI.put(uri, cacheManager);
+        }
+
+        if (!cacheManagersByClassLoader.containsKey(classLoader))
+        {
+            cacheManagersByClassLoader.put(classLoader, cacheManagersByURI);
```

Comment:
Move this into the `if (cacheManagersByURI == null)` block. That way both `cacheManagersByURI` and `cacheManager` read like "try to get it, if it's null, create and add it".